[AIESW-18974] Do not include the scale in LayerNorm recomposing but fuse it later

gerion-amd opened the pull request:
When models have multiple scales, we can match them anyway by providing a scale of 1.
For the normal case, we can fuse it together afterwards.

- The fusing was already in place (just in connection with transpose it needed adaption).
- The operation order after transforming with transpose was different. I adapted the LIT tests accordingly.

## High-Level Explanation:

Previously, only the mul operation was matched that multiplied the scale of the LayerNorm operation. Then, it were matched in three different branches for all possibilities of operations coming before that, so basically:
```c++
struct RecomposeLayerNormFromMulPattern : public OpRewritePattern<ONNXMulOp>{
    LogicalResult matchAndRewrite(ONNXMulOp mulOp, PatternRewriter &rewriter) const final {
        // do the matching
        if (ONNXDivOp precedes mulOp) {
            match_further1();
        } else if(ONNXPowOp precedes mulOp) {
            match_further2();
        } else if(ONNXMulOp precedes mulOp) {
            match_further3();
        } else {
            fail();
        }
        // do the rewrite
    }
};
```
To split out the mulOp (the scale), I converted this to a template with three specializations:
```c++
template<typename DivOperator>
matchOp(DivOperator divOp, remainingParams) {
    fail();
}
template<>
matchOp(ONNXDivOp divOp, remainingParams) {
    match_further1();
}
template<>
matchOp(ONNXPowOp powOp, remainingParams) {
    match_further2();
}
template<>
matchOp(ONNXMulOp mulOp, remainingParams) {
    match_further3();
}

template<typename DivOperator>
struct RecomposeLayerNormFromDivPattern : public OpRewritePattern<DivOperator>{
    LogicalResult matchAndRewrite(DivOperator divOp, PatternRewriter &rewriter) const final {
        matchOp<DivOperator>(divOp, remainingParams);
        // do the rewrite
    }
};

patterns.add<RecomposeLayerNormFromDivPattern<ONNXDivOp>>();
patterns.add<RecomposeLayerNormFromDivPattern<ONNXPowOp>>();
patterns.add<RecomposeLayerNormFromDivPattern<ONNXMulOp>>();
```
The code inside the template specializations is just copied from the previous branches.

gerion-amd commented:
Converted to draft to adapt https://github.com/Xilinx/onnx-mlir/pull/497/files#diff-0d8bdb3d073090c3db042c380fa9ab7168e7c20f9b297f55a52c188c8471ac01R223. This clashes with clang-format and is not nice anyway. Will make a struct out of it.
Edit: Done

jorickert commented on test/mlir/onnx/onnx_recompose_layernorm_with_transpose.mlir:
```diff
@@ -18,15 +19,15 @@ func.func @decomposition_to_layernorm_axis_1(%arg0: tensor<1x4x128x128xf32>) ->
-// CHECK-DAG:       [[VAR_0_:%.+]] = onnx.Constant dense<{{.}}{{.}}[0.0970484465]{{.}}, {{.}}[0.0882187337]{{.}}, {{.}}[0.135120019]{{.}}, {{.}}[0.14907673]{{.}}{{.}}> : tensor<4x1x1xf32>
-// CHECK-DAG:       [[VAR_1_:%.+]] = onnx.Constant dense<{{.}}{{.}}[0.191972837]{{.}}, {{.}}[0.286264896]{{.}}, {{.}}[0.180535644]{{.}}, {{.}}[0.166878015]{{.}}{{.}}> : tensor<4x1x1xf32>
-// CHECK-DAG:       [[VAR_2_:%.+]] = onnx.Constant dense<[1, 4, 1, 1]> : tensor<4xi64>
-// CHECK:           [[VAR_3_:%.+]] = "onnx.Reshape"([[VAR_0_]], [[VAR_2_]]) {allowzero = 0 : si64} : (tensor<4x1x1xf32>, tensor<4xi64>) -> tensor<1x4x1x1xf32>
```
nit: I think if you change this to `CHECK-DAG` you do not need to change anything else in the test

gerion-amd replied:
This worked, thanks! Updated now.

jorickert commented on src/Dialect/ONNX/Transforms/Recompose.cpp:
```diff
@@ -427,13 +506,7 @@ struct RecomposeLayerNormFromMulPattern : public OpRewritePattern<ONNXMulOp> {
       return reportFailure("RMS var eps add has too many uses");
     if (sdSqrtOp && !sdSqrtOp->hasOneUse())
       return reportFailure("RMS std dev sqrt has too many uses");
-    if (powOp && !powOp->hasOneUse())
-      return reportFailure("RMS std dev pow has too many uses");
     // Gate the next 3 ops by being nonnull, as there are multiple paths.
-    if (nDivOp && !nDivOp->hasOneUse())
-      return reportFailure("RMS norm div has too many uses");
-    if (nMulOp && !nMulOp->hasOneUse())
-      return reportFailure("RMS norm mul has too many uses");
```
maybe i missed it. but what happend to these multiple use checks?

gerion-amd replied:
They should not trigger anymore and are kind of moved to the PushTransposeDownScalePattern.
- For the matching of Layernorm without a scale, mulitple uses are allowed now. Fusing of multiple scales is prevented later.
- The transpose pushing is not triggered in case multiple multiplications follow (i.e. more than one scale exist).

gerion-amd committed:
RecomposeLayerNormFromMulPattern: do not match scale

When models have multiple scales, we can match them anyway by
providing a scale of 1.
For the normal case, we can fuse it together afterwards.

jorickert commented on src/Dialect/ONNX/Transforms/Recompose.cpp:
```diff
@@ -677,6 +750,55 @@ struct RecomposeLayerNormFromMulPattern : public OpRewritePattern<ONNXMulOp> {
+    SmallVector<int64_t> perm;
+    ArrayAttrIntVals(*transposePerm, perm);
+
+    scale = create.onnx.upRank(scale, getRank(Y.getType()));
```
nice use of the builder

gerion-amd replied:
AFAIK, it was introduced by @ehsan-toosi exactly for the Layernorm matching with transpose (in this file).

jorickert commented on src/Dialect/ONNX/Transforms/Recompose.cpp:
```diff
@@ -677,6 +750,55 @@ struct RecomposeLayerNormFromMulPattern : public OpRewritePattern<ONNXMulOp> {
+    } else {
+      return rewriter.notifyMatchFailure(mulOp, "no preceding transpose found");
+    }
+    MultiDialectBuilder<OnnxBuilder> create(rewriter, mulOp->getLoc());
```
you are losing the location of the transpose here. Please either fuse it with the mul loc, or add it to the new created transposes

gerion-amd replied:
Good point. I changed that in b5c4d2355 and documented it in b254202dcf.

jorickert commented on src/Dialect/ONNX/Transforms/Recompose.cpp:
```diff
@@ -985,8 +1107,8 @@ struct RecomposeDepthToSpace : public OpRewritePattern<ONNXReshapeOp> {
 
-    // Check for concrete reshape pattern. This pattern applies to both DCR and
-    // CRD modes:
+    // Check for concrete reshape pattern. This pattern applies to both DCR
```
nit: Feel free to ignore: Some unrelated format versions, likely becasue of different clang-format version

gerion-amd replied:
Ok, I saw this too but was to lazy to exclude it out of the commit. I will exclude it next time.

jorickert reviewed (approved):
nice

jorickert commented on src/Dialect/ONNX/Transforms/Recompose.cpp:
```diff
@@ -677,6 +750,55 @@ struct RecomposeLayerNormFromMulPattern : public OpRewritePattern<ONNXMulOp> {
+    // fuseScaleInLayernorm(rewriter, newMulOp, layerNorm)
+    rewriter.replaceOpWithNewOp<ONNXTransposeOp>(
+        mulOp, transposedY.getType(), newMulOp, *transposePerm);
+    return llvm::success();
```
Tiny nit:
```suggestion
    return success();
```

gerion-amd committed:
Layernorm Recompose/PushTransposeDownScalePattern: improve locations

Also reorders the code so the cast to ONNXTransposeOp comes first to
check for a nullptr before using it.

gerion-amd committed:
Layernorm Recompose/PushTransposeDownScalePattern: add documentation

ehsan-toosi commented on test/mlir/onnx/onnx_recompose.mlir:
```diff
@@ -398,6 +399,33 @@ func.func @not_a_layer_norm_with_div_by_two(%input: tensor<1x384x768xf32>, %scal
+// CHECK-DAG:       [[VAR_1_:%.+]] = "onnx.NoValue"() {value} : () -> none
+// CHECK-DAG:       [[VAR_2_:%.+]] = "onnx.Add"([[PARAM_0_]], [[PARAM_1_]]) : (tensor<1x1370x384xf32>, tensor<1x1370x384xf32>) -> tensor<1x1370x384xf32>
+// CHECK:           [[VAR_Y_:%.+]], [[VAR_Mean_:%.+]], [[VAR_InvStdDev_:%.+]] = "onnx.LayerNormalization"([[VAR_2_]], [[VAR_0_]], [[VAR_1_]]) {axis = 2 : si64, epsilon = 3.000000e+00 : f32, stash_type = 1 : si64} : (tensor<1x1370x384xf32>, tensor<384xf32>, none) -> (tensor<1x1370x384xf32>, none, none)
+// CHECK-DAG:       [[VAR_3_:%.+]] = "onnx.Mul"([[VAR_Y_]], [[PARAM_2_]]) : (tensor<1x1370x384xf32>, tensor<384xf32>) -> tensor<1x1370x384xf32>
```
Is there gonna be a follow up PR to fold the mul op into scale of Layernorm?

gerion-amd replied:
This already exists. I haven't search were exactly, but you see in the existing test cases that it happens.

ehsan-toosi commented on src/Dialect/ONNX/Transforms/Recompose.cpp:
```diff
@@ -677,6 +750,79 @@ struct RecomposeLayerNormFromMulPattern : public OpRewritePattern<ONNXMulOp> {
+ *             transpose (loc1)
+ *                 |
+ */
+struct PushTransposeDownScalePattern : public OpRewritePattern<ONNXMulOp> {
```
IMO, this doesn't belong to recompose pass. This is more canonicalization pattern because your goal is to get mul close to the layernorm so that it can fold into scale.

ehsan-toosi commented on src/Dialect/ONNX/Transforms/Recompose.cpp:
```diff
@@ -677,6 +750,79 @@ struct RecomposeLayerNormFromMulPattern : public OpRewritePattern<ONNXMulOp> {
+ *                 |
+ *             transpose (loc1)
+ *                 |
+ */
```
I'd say that only do this if there is a mul with constant and there is a pass that puts constants only on rhs always. Otherwise, you don't want this. So, the case that needs to be supported is this:

```
Layernorm
     |
Transpose    const
      \     /
        mul

```
Then transpose can be folded into const.

gerion-amd replied:
Please correct me, if I'm wrong but isn't it also beneficial to fuse the scale into the Layernorm, even if it's coming not from a const but some other input vector and needs to be transposed?
When not fusing it in this case, a scale is done anyway but with factor 1.

ehsan-toosi replied:
As discussed offline, you are right but I am a bit worried about that extra transpose that is going to be added.

gerion-amd replied:
Make sense, added in 9e27273.

ehsan-toosi commented on src/Dialect/ONNX/Transforms/Recompose.cpp:
```diff
@@ -677,6 +750,79 @@ struct RecomposeLayerNormFromMulPattern : public OpRewritePattern<ONNXMulOp> {
+    // push transpose after mul
+    if (!transposePerm.has_value())
+      return rewriter.notifyMatchFailure(
+          mulOp, "transpose need a fixed permutation");
```
```suggestion
          mulOp, "transpose need a constant permutation");
```

ehsan-toosi commented on src/Dialect/ONNX/Transforms/Recompose.cpp:
```diff
@@ -677,6 +750,79 @@ struct RecomposeLayerNormFromMulPattern : public OpRewritePattern<ONNXMulOp> {
+      return rewriter.notifyMatchFailure(mulOp, "no preceding transpose found");
+    }
+    auto oldTranspose = cast<ONNXTransposeOp>(transposeOp);
+    auto transposePerm = oldTranspose.getPerm();
```
Can't you use onnxShapehelper for transpose to get the perm? you could look at the impl for conv to tosa legalization.

gerion-amd replied:
I'm not really sure how you mean it.

I found a slightly better way to do the permutation (committed in c97e664).
Also, in the same file is an `InputOutputTransposer` class that looked promising (and at least use `inferShapes` for the one transpose operation), but it also needs a manual permutation. I also see no way to use it for operations with multiple inputs (just one mul input needs a transpose in this case).

gerion-amd replied:
I found a way to use it: 6eb9b01fb5207fd53d9e7c08cf4d6c458afdcf29. It should make the transformation more powerful if the permutation is not given directly.

ehsan-toosi replied:
yes, that's what I meant. Thanks.

ehsan-toosi commented on test/mlir/onnx/onnx_recompose.mlir:
```diff
@@ -398,6 +399,33 @@ func.func @not_a_layer_norm_with_div_by_two(%input: tensor<1x384x768xf32>, %scal
 
 // -----
 
+func.func @layernorm_with_double_mul(%arg0: tensor<1x1370x384xf32>, %arg1: tensor<1x1370x384xf32>, %arg2: tensor<384xf32>, %arg3: tensor<384xf32>) -> (tensor<1x1370x384xf32>, tensor<1x1370x384xf32>) {
```
Could you add a bit more tests like that mul is getting folded into Layernorm? One with transpose and one without?

From:
```
Transpose
decomposed layernorm that needs transpose without bias and scale
Transpose
Mul
Add
```

To:
```
Transpose
Layernorm with bias and scale
Transpose
```

gerion-amd replied:
`onnx_recompose_layernorm_with_transpose.mlir` does this test already implicit by having a layernorm with scale and bias that needs to be transposed. The adapted pattern now matches a layernorm without scale and bias just like in your first example while the new pattern converts it to your second example.

However, making it explicit is a good idea. I will add that as tests, too.
Edit: Done

gerion-amd committed:
onnx_recompose_layernorm: add explicit tests for PushTransposeDownScale

gerion-amd committed:
PushTransposeDownScalePattern: move to MulOp Canonicalize

gerion-amd committed:
PushTransposeDOwnScale: use extractFromIntegerArrayAttr

gerion-amd committed:
PushTransposeDownScalePattern: use OnnxTransposeOpShapeHelper

gerion-amd committed:
PushTransposeDownScalePattern: push transpose only if scale is constant

jorickert commented on src/Dialect/ONNX/ONNXOps/Canonicalize.cpp:
```diff
@@ -2434,6 +2434,11 @@ struct PushTransposeDownScalePattern : public OpRewritePattern<ONNXMulOp> {
         return rewriter.notifyMatchFailure(
             mulOp, "transpose without preceding layernorm");
       }
+      auto *op = scale.getDefiningOp();
```
Just for your info, no change needed: 
This can also be written as:
`auto *constOp = scale.getDefiningOp<ONNXConstantOp>();`

Or there is a helper function `isConstLikeValue` that is also able to handle mixed dialect situations

gerion-amd replied:
Thank your for the info. That's good to know!
